### PR TITLE
 commiting forked branch

### PR DESCRIPTION
**Pull request by ovese**
This is the pull request for Gauthier on updated SEIR for irritator

### CODE REVIEW

**Comment by ovese** [comment]
The branch seirdeux has no conflicts. learnt a thing about pull requests from this task.

**Review by quesnel — CHANGES_REQUESTED** [review]
Please, can you change the commit name? For example:
`gui: add QSS1 Seir model`

File: app/gui/node-editor.cpp [hunk 1]
```diff
@@ -2466,6 +2466,28 @@ editor::show_editor() noexcept
                       "Fail to initialize example_qss_izhikevich<1>: %s\n",
                       status_string(ret));
 
+            /*
```
Comment (quesnel) [inline app/gui/node-editor.cpp:2469]: Please, use `clang-format` (available with Visual Studio) to format C++ code.

File: lib/include/irritator/examples.hpp [hunk 2]
```diff
@@ -320,6 +320,260 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
...
+     * @return
+     */
+
+    template<int QssLevel, typename F>
+    status
+        example_qss_seir_lineaire(simulation& sim, F f) noexcept
+    {
+        static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
+
+        bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp:340]: Please, this is an old API. Can you update your repository and use:
`sim.models.can_alloc(10)`

File: lib/include/irritator/examples.hpp [hunk 3]
```diff
@@ -320,6 +320,260 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
...
+    status
+        example_qss_seir_lineaire(simulation& sim, F f) noexcept
+    {
+        static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
+
+        bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&                                    
+            sim.can_alloc<abstract_multiplier<QssLevel>>(2) &&
+            sim.can_alloc<abstract_integrator<QssLevel>>(4) &&
+            sim.can_alloc<constant>(2) &&
+            sim.can_connect(12);//pour le seir_linear qu'est-ce que le nombre de connection 12,
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp:344]: The function `sim.can_connect(12)` checks if the `irt::simulation` object can allocate `12` new connections between models. In irritator simulation kernel (`irritator/core.hpp`), all objects are pre-allocated. No memory allocation is allowed during simulation.

File: lib/include/irritator/examples.hpp [hunk 4]
```diff
@@ -320,6 +320,260 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
...
+        
+
+        return status::success;
+    }
+
+    template<int QssLevel, typename F>
+    status
+        example_qss_seir_nonlineaire(simulation& sim, F f) noexcept
+    {
+        //static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp:416]: You can remove this line.

File: lib/include/irritator/examples.hpp [hunk 5]
```diff
@@ -320,6 +320,260 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
...
+
+        bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&                        
+            sim.can_alloc<abstract_wsum<QssLevel, 3>>(2) &&
+            sim.can_alloc<abstract_multiplier<QssLevel>>(9) &&
+            sim.can_alloc<abstract_integrator<QssLevel>>(4) &&
+            sim.can_alloc<constant>(10) &&
+            sim.can_connect(32);//pour le seir_non_linear qu'est-ce que le nombre de connection 32?
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp:423]: Replace `32` with the correct number of `sim.connect(...)` call and remove the comment.

File: app/gui/node-editor.cpp (hunk 1 above)
Comment (quesnel) [inline app/gui/node-editor.cpp:2469]: Please, use clang-format (available with Visual Studio) to format C++ code.

**Review by quesnel — COMMENTED** [review]
I prefer two patchs. First for examples.hpp, second for the gui.

**Comment by quesnel** [comment]
Some remarks:
- There is many error in the file. For example, `sim.models.can_connect()` can not build. There is no `can_connect()` function in `irt::data_array` class. Prefer the use of  `sim.can_alloc()` and `sim.can_connect()`.
- Use `clang-format` (see: https://devblogs.microsoft.com/cppblog/clangformat-support-in-visual-studio-2017-15-7-preview-1/  for Visual Studio 2019).
- I prefer the rebase instead of merge in git. Use `git rebase origin/master` to merge your code.

File: lib/include/irritator/examples.hpp [hunk 6]
```diff
@@ -25,10 +25,12 @@ example_qss_lotka_volterra(simulation& sim, F f) noexcept
 {
     static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
 
-    bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&
-                   sim.can_alloc<abstract_multiplier<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(2) &&
-                   sim.can_connect(8);
+    // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 7]
```diff
@@ -25,10 +25,12 @@ example_qss_lotka_volterra(simulation& sim, F f) noexcept
...
+    //                sim.can_alloc<abstract_multiplier<QssLevel>>(1) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 8]
```diff
@@ -74,10 +76,12 @@ example_qss_lif(simulation& sim, F f) noexcept
 {
     static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
 
-    bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_cross<QssLevel>>(1) &&
-                   sim.can_alloc<constant>(2) && sim.can_connect(7);
+    // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 9]
```diff
@@ -124,13 +128,15 @@ template<int QssLevel, typename F>
...
 example_qss_izhikevich(simulation& sim, F f) noexcept
 {
-    bool success = sim.can_alloc<constant>(3) &&
-                   sim.can_alloc<abstract_wsum<QssLevel, 2>>(3) &&
-                   sim.can_alloc<abstract_wsum<QssLevel, 4>>(1) &&
-                   sim.can_alloc<abstract_multiplier<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(2) &&
-                   sim.can_alloc<abstract_cross<QssLevel>>(2) &&
-                   sim.can_connect(22);
+    // bool success = sim.can_alloc<constant>(3) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 10]
```diff
@@ -74,10 +76,12 @@ example_qss_lif(simulation& sim, F f) noexcept
 {
     static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
 
-    bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_cross<QssLevel>>(1) &&
-                   sim.can_alloc<constant>(2) && sim.can_connect(7);
+    // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
+    //                sim.can_alloc<abstract_integrator<QssLevel>>(1) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 11]
```diff
@@ -124,13 +128,15 @@ template<int QssLevel, typename F>
...
 {
-    bool success = sim.can_alloc<constant>(3) &&
-                   sim.can_alloc<abstract_wsum<QssLevel, 2>>(3) &&
-                   sim.can_alloc<abstract_wsum<QssLevel, 4>>(1) &&
-                   sim.can_alloc<abstract_multiplier<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(2) &&
-                   sim.can_alloc<abstract_cross<QssLevel>>(2) &&
-                   sim.can_connect(22);
+    // bool success = sim.can_alloc<constant>(3) &&
+    //                sim.can_alloc<abstract_wsum<QssLevel, 2>>(3) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 12]
```diff
@@ -226,10 +232,12 @@ template<int QssLevel, typename F>
 status
 example_qss_van_der_pol(simulation& sim, F f) noexcept
 {
-    bool success = sim.can_alloc<abstract_wsum<QssLevel, 3>>(1) &&
-                   sim.can_alloc<abstract_multiplier<QssLevel>>(2) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(2) &&
-                   sim.can_connect(9);
+    // bool success = sim.can_alloc<abstract_wsum<QssLevel, 3>>(1) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 13]
```diff
@@ -273,10 +281,12 @@ template<int QssLevel, typename F>
 status
 example_qss_negative_lif(simulation& sim, F f) noexcept
 {
-    bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
-                   sim.can_alloc<abstract_integrator<QssLevel>>(1) &&
-                   sim.can_alloc<abstract_cross<QssLevel>>(1) &&
-                   sim.can_alloc<constant>(2) && sim.can_connect(7);
+    // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(1) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 14]
```diff
@@ -337,11 +347,14 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
     {
         static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
 
-        bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&                                    
-            sim.can_alloc<abstract_multiplier<QssLevel>>(2) &&
-            sim.can_alloc<abstract_integrator<QssLevel>>(4) &&
-            sim.can_alloc<constant>(2) &&
-            sim.can_connect(12);//pour le seir_linear qu'est-ce que le nombre de connection 12,
+        // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

File: lib/include/irritator/examples.hpp [hunk 15]
```diff
@@ -415,12 +428,14 @@ example_qss_negative_lif(simulation& sim, F f) noexcept
     {
         //static_assert(1 <= QssLevel && QssLevel <= 3, "Only for Qss1, 2 and 3");
 
-        bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&                        
-            sim.can_alloc<abstract_wsum<QssLevel, 3>>(2) &&
-            sim.can_alloc<abstract_multiplier<QssLevel>>(9) &&
-            sim.can_alloc<abstract_integrator<QssLevel>>(4) &&
-            sim.can_alloc<constant>(10) &&
-            sim.can_connect(32);//pour le seir_non_linear qu'est-ce que le nombre de connection 32?
+        // bool success = sim.can_alloc<abstract_wsum<QssLevel, 2>>(2) &&
```
Comment (quesnel) [inline lib/include/irritator/examples.hpp]: You can remove these comments.

**Review by quesnel — CHANGES_REQUESTED** [review]
It's almost good. Just remove all comments and provides a better commit message. For examples:
examples: add QSS1 SEIR model

**Comment by quesnel** [comment]
There is a conflict in the file `app/gui/node-editor.cpp`. You need to get the latest version with `git rebase origin/master` for example.

**Comment by quesnel** [comment]
I close this pull request: #52 replace #37.

**Pull request closed without merging**